junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/StackTracePruningEngineExecutionListener.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2015-2023 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.launcher.core;
+
+import java.util.function.Predicate;
+
+import org.junit.platform.commons.util.ClassNamePatternFilterUtils;
+import org.junit.platform.commons.util.ExceptionUtils;
+import org.junit.platform.engine.EngineExecutionListener;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestExecutionResult;
+
+/**
+ * Prunes the stack trace in case of a failed event.
+ *
+ * @since 1.10
+ * @see org.junit.platform.commons.util.ExceptionUtils#pruneStackTrace(Throwable, Predicate)
+ */
+class StackTracePruningEngineExecutionListener extends DelegatingEngineExecutionListener {
+
+	public static final String STACKTRACE_PRUNING_ENABLED_PROPERTY_NAME = "junit.platform.stacktrace.pruning.enabled";
+	public static final String STACKTRACE_PRUNING_PATTERN_PROPERTY_NAME = "junit.platform.stacktrace.pruning.pattern";
+	public static final String STACKTRACE_PRUNING_DEFAULT_PATTERN = "org.junit.*,java.*";
+
+	private final String pruningPattern;
+
+	StackTracePruningEngineExecutionListener(EngineExecutionListener delegate, String pruningPattern) {
+		super(delegate);
+		this.pruningPattern = pruningPattern;
+	}
+
+	@Override
+	public void executionFinished(TestDescriptor testDescriptor, TestExecutionResult testExecutionResult) {
+		if (testExecutionResult.getThrowable().isPresent()) {
+			Throwable throwable = testExecutionResult.getThrowable().get();
+
+			ExceptionUtils.findNestedThrowables(throwable).forEach(this::pruneStackTrace);
+		}
+		super.executionFinished(testDescriptor, testExecutionResult);
+	}
+
+	private void pruneStackTrace(Throwable throwable) {
+		ExceptionUtils.pruneStackTrace(throwable,
+			ClassNamePatternFilterUtils.excludeMatchingClassNames(pruningPattern));
```

Comment:
I think we should create the `Predicate` in the constructor so it can be reused here rather than compiling `Patterns` for every `Throwable`.